chore!: NEXT-40590 - Fix discovery deprecations 2 (Media, CMS Storefront, SalesChannel)

MartinKrzykawski opened the pull request:
<!--
Thank you for contributing to Shopware! Please fill out this description template to help us to process your pull request.

Please make sure to fulfill our contribution guideline (https://developer.shopware.com/docs/resources/guidelines/code/contribution?category=shopware-platform-dev-en/contribution).

Do your changes need to be mentioned in the documentation?
Please create a second pull request at https://github.com/shopware/docs
-->

### 1. Why is this change necessary?
Have deprecated code which needs to be fixed for 6.7.


### 2. What does this change do, exactly?
Fixed the deprecations for:
* Media
* CMS Storefront templates
* SalesChannel

I deleted wrong deprecations in sw-cms-list component because these functions are still required for the prod code and should not be deleted.

### 3. Describe each step to reproduce the issue or behaviour.


### 4. Please link to the relevant issues (if any).
<!-- Examples:
- closes #123  - closes the issue #123 when the PR is merged
- relates #123 - relates to the issue #123

In case of issue existing only on Jira, link to the Jira issue.
- Jira issue: https://shopware.atlassian.net/browse/NEXT-123
-->

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [ ] I have read the contribution requirements and fulfill them.

MartinKrzykawski committed:
NEXT-40590 - Fix media deprecations

MartinKrzykawski committed:
NEXT-40590 - Fixed storefront cms deprecations

mitelg commented on src/Storefront/Resources/views/storefront/element/cms-element-category-navigation.html.twig:
```diff
@@ -1,21 +1,16 @@
-{# @deprecated tag:v6.7.0 - Variable `active` will be removed, as it will be unused #}
```
I already took care of this file here https://github.com/shopware/shopware/pull/6575

MartinKrzykawski replied:
i reverted the changes from my git history, so we will not run into conflicts 👍

MartinKrzykawski committed:
NEXT-40590 - Fixed sales-channel deprecations

MartinKrzykawski committed:
NEXT-40590 - Fixed phpstan

mitelg commented on src/Core/System/SalesChannel/Aggregate/SalesChannelType/SalesChannelTypeEntity.php:
```diff
@@ -138,11 +96,25 @@ public function setIconName(string $iconName): void
+     *
+     * @deprecated tag:v6.8.0 - reason:parameter-type-change - parameter type will be nullable
+     */
     public function setScreenshotUrls(array $screenshotUrls): void
```
widening the allowed parameter is not a breaking change. so you can just change it.

MartinKrzykawski replied:
oh you are right 💡 thanks man, thats perfect!

mitelg replied:
now you didn't changed the parameter, but widened the return type of the getter instead :grimacing: 

that is probably a break, as before it was sure, that always an array was returned, but it could be null now. 

while allowing an additional parameter in the setter is not a break

MartinKrzykawski committed:
NEXT-40590 - Fixed method signatures in SalesChannelTypeEntity, added deprecations and wrote changelog

MartinKrzykawski committed:
NEXT-40590 - Removed wrong deprecations in sw-cms-list component

marcelbrode commented on src/Administration/Resources/app/administration/src/module/sw-cms/page/sw-cms-list/index.js:
```diff
@@ -580,23 +580,14 @@ export default {
-        /**
-         * @deprecated tag:v6.7.0 - Will be removed
-         */
         getPageCategoryCount(page) {
             return page.categories.length;
         },
```
Just curiosity: Why no remove? Do we still need this?

MartinKrzykawski replied:
yes, still required and used

marcelbrode commented on src/Core/Content/Media/MediaType/MediaType.php:
```diff
@@ -8,19 +8,12 @@
+    protected string $name;
 
     /**
      * @var array<string>
```
Wouldn't `list<string>` better in that case?

MartinKrzykawski replied:
no because this would result in a list of string values but the flags property allows mixed indices

marcelbrode replied:
then, how about `array<mixed, string>`?

mitelg replied:
an array key can only be a string or integer :wink:

with `array<string>` you already omit the information about the key type, so it can already be both

bschulzebaek commented on src/Storefront/Resources/views/storefront/element/cms-element-form/form-components/cms-element-form-textarea.html.twig:
```diff
@@ -1,47 +1,16 @@
+        additionalClass: additionalClass,
+    } %}
 
     {% else %}
```
We don't need these 2 lines anymore, right?`

marcelbrode commented on src/Core/System/SalesChannel/Event/SalesChannelContextPermissionsChangedEvent.php:
```diff
@@ -11,12 +11,7 @@
-     * @deprecated tag:v6.7.0 - Will be natively typed
-     */
-    protected $permissions = [];
+    protected array $permissions = [];
```
array annotation missing

MartinKrzykawski replied:
i checked the code and it might be that this array is from type `array<string, bool>` but i kept it like before because i was not sure if this is not a break since this property did not have any specification before

MartinKrzykawski replied:
@mitelg any opinion here?

marcelbrode commented on src/Core/System/SalesChannel/SalesChannelEntity.php:
```diff
@@ -48,516 +48,163 @@ class SalesChannelEntity extends Entity
+    protected int $navigationCategoryDepth;
 
     /**
      * @var array<mixed>|null
```
```suggestion
     * @var array<string, mixed>|null
```

Index type would be great

MartinKrzykawski replied:
would it not be a break to change a type declaration without deprecating it before and how do you know that the index has to be a string? i do not find any reference for this

marcelbrode replied:
Annotations are more like browser/static anaylze infos and are not breaking. Would be different, with native typing.

In this case I just happen to know that. SlotConfig always looks like this:
```
                'slotConfig' => [
                    $slotId => [
                        'content' => [
                            'source' => 'static',
                            'value' => $languageCode . ' Override',
                        ],
                        'maybeSomethingDifferent' => 'something of mixed type',
                    ],
                ],
```

marcelbrode commented on src/Core/System/SalesChannel/SalesChannelEntity.php:
```diff
@@ -48,516 +48,163 @@ class SalesChannelEntity extends Entity
+    protected ?MainCategoryCollection $mainCategories = null;
 
     /**
      * @var array<string>|null
```
```suggestion
     * @var list<string>|null
```
Maybe?

bschulzebaek commented on src/Storefront/Resources/views/storefront/element/cms-element-form/form-types/newsletter-form.html.twig:
```diff
@@ -2,42 +2,38 @@
+                    'data-form-field-toggle-value': 'subscribe',
+                },
+            } %}
                 <div class="form-group col-12">
```
The following `form-group` is the deprecated part and should be removed

marcelbrode commented on src/Storefront/Resources/views/storefront/element/cms-element-form/form-types/newsletter-form.html.twig:
```diff
@@ -2,42 +2,38 @@
-                            'data-form-field-toggle-value': 'subscribe',
-                        },
-                    } %}
-                {% else %}
```
The else case is gone, but the code is still there. Probably has to be removed as well.

marcelbrode reviewed (changes requested):
Nice job so far 👍

mitelg commented on src/Core/System/SalesChannel/Aggregate/SalesChannelType/SalesChannelTypeEntity.php:
```diff
@@ -118,32 +76,58 @@ public function setDescriptionLong(?string $descriptionLong): void
+     */
     public function getCoverUrl(): string
     {
+        if ($this->coverUrl === null) {
```
maybe it would be good, to add a hint, that this ifs can be removed with 6.8. I am not sure, if we will remember in a year, what to do here :sweat_smile:

MartinKrzykawski replied:
good point, i will update the deprecations 👍

MartinKrzykawski committed:
NEXT-40590 - Code review fixes

MartinKrzykawski committed:
Merge branch 'trunk' into next-40590/fix-discovery-deprecations-part-2

MartinKrzykawski committed:
NEXT-40590 - Reverted the setter functions setTranslations and setSalesChannels for SalesChannelTypeEntity because they are not allowed to be nullable as they are toMany associations

MartinKrzykawski commented:
FYI: the pipeline failed because setter functions where the property is defined as a toMany association in the definition class, cannot be nullable. I reverted the change for the SalesChannelTypeEntity for the functions `setSalesChannels` and `setTranslations`. Thats the only thing that i changed. I enabled the auto-merge :)

mitelg commented:
> ... because setter functions where the property is defined as a toMany association in the definition class, cannot be nullable.

interesting and good to know :see_no_evil: where is that defined?

MartinKrzykawski commented:
> > ... because setter functions where the property is defined as a toMany association in the definition class, cannot be nullable.
> 
> interesting and good to know 🙈 where is that defined?

Totally! :D It is defined in `Shopware\Core\Framework\DataAbstractionLayer::validateSetterIsNotNull` - it is part of the dal:validate process

MartinKrzykawski committed:
NEXT-40590 - Changed media property to nullable and deprecated the getter method

MartinKrzykawski commented:
Hey @mitelg, i would be happy about your opinion to my last change:

I changed the property `media` in the `MediaFolderEntity` to make it nullable.
The phpdoc annotation was defined as non-nullable before, so this could be wrong but i have some reasons for this:

The property is a association property and should not require to be initialized before imo. 
I found also in the `MediaFolderDefinition` that the referenced field is set to null if the media reference is deleted:
`(new OneToManyAssociationField('media', MediaDefinition::class, 'media_folder_id'))->addFlags(new SetNullOnDelete())`.

My most important reason is that i run into php errors if i tried to create a new media folder in the administration with a required media property:

![mediapropissue](https://github.com/user-attachments/assets/64ca66e6-0469-4e25-b55d-144f0cf914f6)

In my opinion this property should be nullable to behave like before and to not break other places.

I noticed this wrong behavior because the `JsonEntityEncoderTest` and the `JsonApiEncoderTest` failed in the integration pipeline. Now they behave like before with the only addition that i had to update the testing fixtures to require a media name since this property is required now.

I will not merge this before your okay because it might be that you have a different opinion :)

mitelg commented:
@MartinKrzykawski totally fine to make this property nullable. I guess, it was just missed in the annotation.

MartinKrzykawski committed:
Merge branch 'trunk' into next-40590/fix-discovery-deprecations-part-2

MartinKrzykawski committed:
Merge branch 'trunk' into next-40590/fix-discovery-deprecations-part-2

MartinKrzykawski committed:
Merge branch 'trunk' into next-40590/fix-discovery-deprecations-part-2

MartinKrzykawski committed:
Merge branch 'trunk' into next-40590/fix-discovery-deprecations-part-2
